Rebase for feature #1991 to branch dev-1.21

geekwisdom opened the pull request:
This is the rebase for #1991 in branch dev-1.2.1

==
This is the pull request for #1991. It implements a new 'dogequirks' class that can be extended to specific doge coin items without (hopefully) to much disturbance to keeping aligned upstream.

I am not 100% sure I got the GetMinFee procedure correct. If I understand it correctly it halves on each call, so I may not have put the moon_phase code in exactly the 'right' spot, it's a little hard to test.

Perhaps it should also be (or only be) in miner.cpp

```
if (IsArgSet("-blockmintxfee")) {
        CAmount n = 0;
        ParseMoney(GetArg("-blockmintxfee", ""), n);
        blockMinFeeRate = CFeeRate(n);
    } else {
        blockMinFeeRate = CFeeRate(DEFAULT_BLOCK_MIN_TX_FEE);
        **// Add moon check code here??**
         
    }


```
Appropriate unit test class has been added as well, the specific unit test works but I re-ran all (now 254) unit tests and many failed which seem to have nothing to do with my change, some unit test clean up maybe required?

Required Testing: / Test Plan

a) Verify the calculation of the transaction fee calculated on one of the '3' days of 'full moon' show at approx. half the rate as the other days of the month

b) Try changing the constant "PROPOSED_TX_FEE" to a non-zero, in mem-pool.h. Does the tx fee then become the proposed fee when preforming a calculation

c) What happens if the minrate is really low, and then gets halved, will the function still work correctly with very small 0.000000001 fees?

patricklodder commented:
Removed "consensus", because fees aren't subject to that.

TheHolyRoger commented:
Any thoughts about what will happen when people using these cheaper moon phase fees wonder why their transactions aren't confirming as quick and the mempool ramps up?

Ivegotnofloor commented:
> Any thoughts about what will happen when people using these cheaper moon phase fees wonder why their transactions aren't confirming as quick and the mempool ramps up?

What if its cloudy outside.

langerhans commented:
Question: How do you make sure the nodes are on the same time? Imagine this: I create a transaction with a low fee cause my computer time is behind and it's still "moon" for me. Now the transaction is broadcast but gets rejected cause for the miners "moon" is over already. 
There is no common source of truth for time like this, is there?

geekwisdom commented:
> Any thoughts about what will happen when people using these cheaper moon phase fees wonder why their transactions aren't confirming as quick and the mempool ramps up?

If miners wait out the moon phase this would be a signal they dont support the fee reduction meaning there should not be an umcoming fee change. If the transactions continue to confirm this would (I think) indicate they see the value to everyone in the chain.  

The reduction during moon phases is easily deductable

geekwisdom commented:
> Question: How do you make sure the nodes are on the same time? Imagine this: I create a transaction with a low fee cause my computer time is behind and it's still "moon" for me. Now the transaction is broadcast but gets rejected cause for the miners "moon" is over already.
> There is no common source of truth for time like this, is there?

The moon phase is calculate in utc by date only (not time). Time drift is managed by the timeoffset built into the protocol. Miners and nodes should be syncing with ntp or similar anyway to keep the drift as small as possible 

See: https://bitcoin.stackexchange.com/questions/43638/bitcoind-daemon-return-timeoffset-10-15-50-60-and-more 

Bitcoin but applies to dogecoin as well

rnicoll commented:
I think using most recent block time to indicate time would be stabler, as there's no dependency on the server's local clock being correct. I acknowledge that miner's clocks have to fairly correct and the nodes accepting blocks have to have vaguely sane clocks, but still somewhat safer.

Question, though; what happens if a transaction is raised in one moon phase, but doesn't get included in a block in time, can there be a month-long pause before the transaction is mineable? I'm wondering if rather than minfee, this should adapt the fee recommendation in the UI slightly?

Secondary concerns around non-reference implementations of fees, as many wallets use standard Bitcoin (TX size x an estimate from recent transactions) fee calculation logic and would miss the nuance of this. Again why I think maybe this is something that's a minor aesthetic effect rather than adjusting minimums.

geekwisdom commented:
> I think using most recent block time to indicate time would be stabler, as there's no dependency on the server's local clock being correct. I acknowledge that miner's clocks have to fairly correct and the nodes accepting blocks have to have vaguely sane clocks, but still somewhat safer.
> 
> Question, though; what happens if a transaction is raised in one moon phase, but doesn't get included in a block in time, can there be a month-long pause before the transaction is mineable? I'm wondering if rather than minfee, this should adapt the fee recommendation in the UI slightly?
> 
> Secondary concerns around non-reference implementations of fees, as many wallets use standard Bitcoin (TX size x an estimate from recent transactions) fee calculation logic and would miss the nuance of this. Again why I think maybe this is something that's a minor aesthetic effect rather than adjusting minimums.

Still learning my way around the code :-)  No idea what the effect would be when raised in moon phase but not in bloc k time. I was thinking exactly the opposite, and that by putting adjusting the minfee at mempool, a worried user could still choose to push the transaction with a higher fee if they were worried about it getting stuck, BUT by putting it in the GUI "recommended", the user might think the 'recommended' more of a 'guarantee' of a fast transaction' when it fact it might actually be slower. (or not work at all)

My suggestion would be for users to get a discount on fees at specific times, but the 'How', I am uncertain off. I would not want users transactions to get 'stuck', nor do I want the transactions to take months.

But perhaps I am simply misunderstanding how it works, I will search around in more of the UI to adjust the recommended fee setting (though if anyone knows where that is, and could save me the time hunting, that would be nice too :-))

For now I have updated PR, removed calc from mempool and also updated the code to take in directly as int64_t nTime instead of the weird scansf stuff I just did as a quick hack to try and get something 'working'

WarlockD commented:
I like this idea.  Coming from the meme heavy SS13 codebase, "More memes are good memes"  It might also help with other problems like scheduling transactions during times when the fee can be higher and lower volume, if you want to go practical with it.  But then I do like the idea of making a time called "420" and the fee be a multiplier 0.420 that hah.   I am going to dig well this seems fun.

michilumin commented:
Guys I'm looking into doing **something** with this in 1.21; at least a 'moon widget' or something that may vary the recommended fee set. Having it be required is a longer pole as we run into issues with, as you mentioned, time truth -  as well as non-core wallets likely not wanting to implement the 'moon phase thing'.

But it would be nice to at least get a nod to this implemented, and we can flesh it out later.

fmhc commented:
What about using this in reverse and having an optional but preselected "moon fee" (old tx fee amount of 1 doge) on full moons in the 1.21 wallet to add Doge donation to a charity / to the dev community wallet, a "round up" function? could even make sense to include sth. like a slider in the UI, so ppl can choose if low tx or donation tx?

AbcSxyZ commented:
@fmhc Give to charity sounds good :) Perhaps adding an output is easier than reducing transaction fees ?
What about doing it the entire month, changing every full moon, community selecting (vote?) each month a new cause :open_mouth: 

Charity inside the Dogecoin core !

bblackbeard commented:
> @fmhc Give to charity sounds cool :) Perhaps adding an output is easier than reducing transaction fees ?
> What about doing it the entire month, changing every full moon, community selecting (vote?) each month a new cause 😮
> 
> Charity inside the Dogecoin core, doing only good every day 😉

Climate change  charity donations(hence the big issue is "ECO Friendly" for crypto, save the planet (tree projects, reduce carbon projects)
